packages/@lwc/engine-dom/src/patches/global-registry.ts

```diff
@@ -0,0 +1,398 @@
+            );
+        }
+        definitionsByTag.set(tagName, definition);
+        definitionsByClass.set(constructor, definition);
+        let PivotCtor = pivotCtorByTag.get(tagName);
+        if (isUndefined(PivotCtor)) {
+            PivotCtor = createPivotingClass(tagName, definition);
+            // Register a pivoting class which will handle global registry initializations
+            nativeDefine.call(nativeRegistry, tagName, PivotCtor);
+        }
+        // For globally defined custom elements, the definition associated
+        // to the UserCtor has a back-pointer to PivotCtor in case the user
+        // new the UserCtor, so we know how to create the underlying element.
+        definition.PivotCtor = PivotCtor;
+        // Upgrade any elements created in this scope before define was called
+        // which should be exhibit by LWC using a tagName (in a template)
+        // before the same tagName is registered as a global, while others
+        // are already created and waiting in the global context, that will
+        // require immediate upgrade when the new global tagName is defined.
+        const awaiting = awaitingUpgrade.get(tagName);
+        if (!isUndefined(awaiting)) {
+            awaitingUpgrade.delete(tagName);
+            for (const element of awaiting) {
+                const registeredDefinition = pendingRegistryForElement.get(element);
+                if (!isUndefined(registeredDefinition)) {
+                    pendingRegistryForElement.delete(element);
+                    internalUpgrade(element, registeredDefinition, definition);
+                }
+            }
+        }
+        // Flush whenDefined callbacks
+        const resolver = definedResolvers.get(tagName);
+        if (resolver) {
+            resolver(constructor);
+        }
+    };
+
+    CustomElementRegistry.prototype.get = function get(
+        this: CustomElementRegistry,
+        tagName: string
+    ): CustomElementConstructor | undefined {
+        return nativeGet.call(this, tagName) && definitionsByTag.get(tagName)?.UserCtor;
+    };
+
+    CustomElementRegistry.prototype.whenDefined = function whenDefined(
+        this: CustomElementRegistry,
+        tagName: string
+    ): Promise<CustomElementConstructor> {
+        return nativeWhenDefined.call(this, tagName).then(() => {
+            let promise = definedPromises.get(tagName);
+            if (isUndefined(promise)) {
+                const definition = definitionsByTag.get(tagName);
+                if (!isUndefined(definition)) {
+                    return definition.UserCtor;
+                }
+                let resolve: (constructor: CustomElementConstructor) => void;
+                promise = new Promise((r) => (resolve = r));
+                definedPromises.set(tagName, promise);
+                definedResolvers.set(tagName, (ctor) => resolve(ctor));
+            }
+            return promise;
```

Comment:
@caridy I'm not able to think of a case where we need `definedPromises` or `definedResolvers`. If we're in the `then()` callback for a `nativeWhenDefined` call, then by definition we've already set the definition. I think we can just simplify this to always return `definition.UserCtor`.